### PR TITLE
fix: minBrowserVersion in SyncOption should be a number

### PR DESCRIPTION
**Pull request by rerorero**
## Issue to fix
This PR fixes an issue where selectMatchingDrivers() ignores `minBroswerVersion` and attempts to download a lot of unnecessary ChromeDrviers.

[Here is a log](https://github.com/appium/appium-uiautomator2-driver/suites/11504452138/artifacts/594646158). While it can detect the Chrome version, it doesn't filter with the minimum browser version. 
```
...... (At Line 12996)
2023-03-12 08:12:37:097 - [debug] [Chromedriver@558d] [Chromedriver@558d] Browser version in the supplied details: Chrome/66.0.3359.158
2023-03-12 08:12:37:097 - [debug] [Chromedriver@558d] [Chromedriver@558d] Found Chrome bundle 'undefined' version '66.0.3359'
...
... (At Line 16182)
2023-03-12 08:12:41:829 - [debug] [ChromedriverStorageClient] [ChromedriverStorageClient] Selecting chromedrivers whose platform matches to mac64
2023-03-12 08:12:41:831 - [debug] [ChromedriverStorageClient] [ChromedriverStorageClient] Got 121 items
...
```

## Cause
[selectMatchingDrivers() assumes `minBrowserVersion` is a string](https://github.com/appium/appium-chromedriver/blob/e4e2a82664896fb654ceaa89dce64267871752b9/lib/storage-client.js#L299) but it's always a number as it's [given here](https://github.com/appium/appium-chromedriver/blob/25b1c64e6f70f1138f0a0e4ddffffb9ecae5575d/lib/chromedriver.js#L350) and the type of `Semver.chromeVersion.major` is number. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/semver/classes/semver.d.ts#L12

### CODE REVIEW

File: lib/storage-client.js [hunk 1]
```diff
@@ -296,11 +296,10 @@ class ChromedriverStorageClient {
       }
     }
 
-    if (_.isString(minBrowserVersion) && !Number.isNaN(minBrowserVersion)) {
+    if (minBrowserVersion && !Number.isNaN(minBrowserVersion)) {
       // Only select drivers that support the current browser whose major version number equals to `minBrowserVersion`
-      const minBrowserVersionInt = parseInt(minBrowserVersion, 10);
```
Comment (mykola-mokhnach) [inline lib/storage-client.js:301]: I think we could keep this cast just to be on the safe side

Comment (rerorero) [reply lib/storage-client.js:301]: Sure, then let me revert the type to `string | number`

File: lib/storage-client.js [hunk 2]
```diff
@@ -296,11 +296,10 @@ class ChromedriverStorageClient {
       }
     }
 
-    if (_.isString(minBrowserVersion) && !Number.isNaN(minBrowserVersion)) {
+    if (minBrowserVersion && !Number.isNaN(minBrowserVersion)) {
```
Comment (mykola-mokhnach) [inline lib/storage-client.js:299]: this condition could be changed to `/^\d+$/.test(minBrowserVersion)`

Comment (rerorero) [reply lib/storage-client.js]: I think we need a typecast and null check like this https://github.com/appium/appium-chromedriver/pull/291/commits/543d94868bd46836df9a1554b0896e8dc0ad1b08 for typescript linter.

File: lib/utils.js [hunk 3]
```diff
@@ -147,6 +147,27 @@ function generateLogPrefix(obj, sessionId = null) {
...
+/**
+ * Converts the given object to an integer number if possible
+ *
+ * @param {any} value to be converted
+ * @returns {number?}
+ */
+function convertToInt(value) {
+  switch (typeof value) {
+    case 'number':
+      return value;
```
Comment (mykola-mokhnach) [inline lib/utils.js:159]: typeof NaN === 'number' in javascript ;)

File: lib/storage-client.js [hunk 4]
```diff
@@ -296,9 +297,9 @@ class ChromedriverStorageClient {
       }
     }
 
-    if (_.isString(minBrowserVersion) && !Number.isNaN(minBrowserVersion)) {
+    const minBrowserVersionInt = convertToInt(minBrowserVersion);
+    if (minBrowserVersionInt) {
```
Comment (mykola-mokhnach) [inline lib/storage-client.js:301]: do we also need to reject zeroes?

File: lib/utils.js (hunk 3 above)
Comment (rerorero) [reply lib/utils.js:159]: Thanks, I'll fix it and add a test

File: lib/storage-client.js (hunk 4 above)
Comment (rerorero) [reply lib/storage-client.js:301]: I think zero should be valid technically.  let me fix it to validate with null.

Comment (rerorero) [reply lib/storage-client.js]: fixed https://github.com/appium/appium-chromedriver/pull/291/commits/b90aecc025e87d9310f1adc8b6f2b05fdf1173dc

File: lib/utils.js (hunk 3 above)
Comment (rerorero) [reply lib/utils.js:162]: fixed https://github.com/appium/appium-chromedriver/pull/291/commits/6505403a5cb7c03fac55ea2e6f73b3de17108191

File: lib/utils.js [hunk 5]
```diff
@@ -147,6 +147,30 @@ function generateLogPrefix(obj, sessionId = null) {
...
+/**
+ * Converts the given object to an integer number if possible
+ *
+ * @param {any} value to be converted
+ * @returns {number?}
+ */
+function convertToInt(value) {
+  switch (typeof value) {
+    case 'number':
+      if (Number.isNaN(value)) {
```
Comment (mykola-mokhnach) [inline lib/utils.js:159]: this and the below conditions could be simplified using the ternary operator, for example: `return Number.isNaN(value) ? null : value;`

File: lib/utils.js [hunk 6]
```diff
@@ -147,6 +147,30 @@ function generateLogPrefix(obj, sessionId = null) {
...
+      return value;
+    case 'string': {
+      const parsedAsInt = parseInt(value, 10);
+      if (!Number.isNaN(parsedAsInt)) {
+        return parsedAsInt;
+      }
+      return null;
+    }
+  }
+  return null;
```
Comment (mykola-mokhnach) [inline lib/utils.js:171]: it is usually a better style to put return into the `default` section

File: lib/utils.js [hunk 7]
```diff
@@ -147,6 +147,30 @@ function generateLogPrefix(obj, sessionId = null) {
   );
 }
 
+/**
+ * Converts the given object to an integer number if possible
+ *
+ * @param {any} value to be converted
+ * @returns {number?}
```
Comment (mykola-mokhnach) [inline lib/utils.js:154]: could also be `{number | null}`

**Comment by mykola-mokhnach** [comment]
It looks like unit tests need some update

**Comment by rerorero** [comment]
Thank you for reviews,  I'll take a look at CI not working.

**Comment by rerorero** [comment]
I fixed the points according to your comments, PTAL
And I found the CI issue doesn't relate to this change. So I opened a new PR, please also take a look. Thank you.
https://github.com/appium/appium-chromedriver/pull/292

File: lib/utils.js [hunk 8]
```diff
@@ -147,6 +147,25 @@ function generateLogPrefix(obj, sessionId = null) {
...
+/**
+ * Converts the given object to an integer number if possible
+ *
+ * @param {any} value to be converted
+ * @returns {number | null}
+ */
+function convertToInt(value) {
+  switch (typeof value) {
+    case 'number':
+      return Number.isNaN(value) ? null : value;
```
Comment (mykola-mokhnach) [inline lib/utils.js:159]: this might also return floats, but that is probably not something we care much

**Pull request merged**